### ISSUE TITLE
[SYS-35] The semantic filter permanently rejects 30-55% of articles from every outlet's dedicated finance section (world oil prices 0.254, Boeing shares 0.203, BI/QRIS 0.288): the raw threshold has no recall floor and the category gate only works in the deny direction

### ISSUE THREAD

**Issue by imanueldrexel** [imanueldrexel/news-scraping#118]
- **Where:** `newscrawler/core/importance_filter.py:125-162` (`SemanticImportanceFilter.is_important` -- raw cosine threshold 0.35 against the short `INTERESTS` phrases, applied to every article regardless of category); `newscrawler/domain/services/crawler_service_impl.py:161-175` (a rejection becomes `sitemaps.crawl_status='rejected_filter'`, SYS-16's permanent terminal state).
- **Symptom:** on outlets' *dedicated finance sections* -- where the category alone already says "finance" -- the semantic filter permanently rejects **30-55% of articles**: LIPUTAN6 `bisnis` 111 rejected / 89 saved (55.5%), SUARA `bisnis` 47.0%, MERDEKA `uang` 44.0%, WARTAEKONOMI `bisnis` 44.4%, IDNTIMES `business` 40.7%, SINDONEWS `ekbis` 39.0%, TEMPO `ekonomi` 37.0%, MEDIAINDONESIA `ekonomi` 35.9%, JPNN `ekonomi` 36.0%, TIRTO `ekonomi` 33.5%, TVONENEWS `ekonomi` 29.5%; VIVA `bisnis` 42% in a 150-row dry run. The rejected rows are not noise. A random 36 from this week's `rejected_filter` set on those sections includes: `Harga Minyak Dunia Turun Usai AS Ungkap Pipa Arab Saudi Segera Beroperasi`, `Produksi Pesawat Terhambat, Saham Boeing Turun Tajam`, `BI Jajaki Kerja Sama QRIS dengan 4 Negara Baru`, `Utang Whoosh Akan Dilimpahkan Ke Kemenkeu Pertengahan September`, `Harga Minyak Dunia Turun Tipis Dipicu Kebijakan Arab Saudi`, `DJP Gandeng Pertamina Perkuat Sistem Perpajakan`, `Pintu Rilis Layanan VIP Private, Khusus Trader Aset Rp10 Miliar`, `Bahlil: Perusahaan Rusia Mau Bangun Pabrik Alumina di Kalimantan`, `Rokok Ilegal Marak, Realisasi Pajak Rokok Jakarta Turun Rp37 M`, `Maruarar Sirait Targetkan Akad Massal 71 Ribu Rumah Subsidi`. ~30 of the 36 are genuine economy/market news; the pipeline's stated interests include `"Pasar Saham"`, `"Berita Penggerak Pasar"`, `"Komoditas"`, `"Kebijakan Bank Sentral"`.
- **Root cause:** it is the model, not the extractor. Re-fetching 8 of those rows through the real crawler gives `extractor=site`, clean bodies (`"Liputan6.com, Jakarta - Harga minyak dunia turun pada perdagangan Rabu... Harga minyak mentah West..."`), and the real `check_text` (title + body[:500]) scores **0.254** against its best interest (`Berita Penggerak Pasar`); the Boeing article 0.203 (`Laporan Keuangan Perusahaan`); the BI/QRIS article 0.288 (`Kebijakan Bank Sentral`); Whoosh debt 0.344; DJP/Pertamina tax 0.332; Pintu crypto 0.322 (`Pasar Saham`); Rusal alumina 0.328. Cosine similarity between a `paraphrase-multilingual-MiniLM-L12-v2` article embedding and one-to-three-word interest phrases is too weak a signal for broad economy coverage; it only works on narrowly-financial text (WARTAEKONOMI `perbankan`/`makro`/`bursa`/`finansial` and CNBC `market` reject just 6-12%). Adding the body makes it *worse* in 6 of the 8 cases (title-only 0.341 -> title+body 0.203 for Boeing). SYS-31 (#86) fixed precision at the top of the score range and left recall at the bottom untouched -- the audit that closed it read the 45-65% pass rates on the newly-fixed outlets as "a healthy finance-focused feed" without reading the rejects. SYS-31's category gate is also asymmetric: a known-off-topic category short-circuits to reject, but a known-finance category still has to clear the same threshold as a `news` bucket. And since SYS-16 (#37), every one of these false negatives is `rejected_filter` -- never re-queued, never retried.
- **Evidence:** `SELECT sources, category, COUNT(*) FILTER (WHERE crawl_status='rejected_filter'), COUNT(a.sitemap_id) FROM sitemaps s LEFT JOIN articles a USING (sitemap_id) WHERE lower(category) IN ('bisnis','ekonomi','uang','business','ekbis','market',...) GROUP BY 1,2` -- table above; ~1,100 rows permanently rejected under finance categories, ~800 of them stamped this week (2026-09-19/20 runs). Dry-run attribution on 150 VIVA `bisnis` rows through the real crawler and filter: kept 83, below-threshold 63, margin-rejected 4 -- the loss is the threshold, not SYS-31's margin. Re-scoring script and the 8 worked examples are in the 2026-09-20 VIVA entry in `EXPERIMENTS.md`.
- **Fix:** (1) make the category gate symmetric: a `FINANCE_CATEGORIES` allow-set (`bisnis`, `ekonomi`, `uang`, `finance`, `business`, `ekbis`, `saham`, `keuangan`, `makro`, `bursa`, `finansial`, `perbankan`, `market`, `Ekonomi`/`Finansial`/`Bisnis`/`Bursa` from ANTARANEWS, ...) that short-circuits `is_important` to `True` *before* the semantic score, exactly mirroring `is_offtopic_category` -- every outlet fixed this week is finance-section-only, so on those the semantic score is only ever discarding real content; the score then arbitrates only the genuinely mixed buckets (`news`, `berita`, `news-sitemap`, regional subdomains), which is what it was built for. (2) Reset every `rejected_filter` row whose category is in that allow-set back to `pending` so the cooldown re-offers it (the body was never saved, so a re-fetch is required; ~1,100 rows). (3) Longer term, or instead of (1) if the allow-list is rejected on design grounds: replace the phrase anchors with a few sentence-length descriptions per interest (embeddings of sentences match sentences) and re-tune the threshold on a labelled sample that includes *rejects*, not just saves -- `probe_filter_precision.py` only ever measured what was saved. (4) Add a recall guard: a real-model test that the eight worked examples above (title + real body head) pass, and a `check_crawlers.py`-style per-run stat of pass rate on allow-set categories so a future model/threshold change that drops it below ~90% is visible.
- **Done when:** the eight worked examples pass `is_important` with their real bodies; a real `task=full_text` over a finance-section backlog (LIPUTAN6 `bisnis` or VIVA `bisnis`) saves >= 90% of extracted rows; the ~1,100 finance-category `rejected_filter` rows are re-queued and re-crawled; the pass-rate stat exists.

Found by the 2026-09-20 VIVA audit, then confirmed on every finance-only outlet crawled this week. Not a regression of SYS-31 (#86) -- SYS-31 is correct about precision; this is the recall side it never measured. Related: SYS-16 (#37) makes the loss permanent; DBT-05 (#24); the "Pass-through rates (45-65%) look like a healthy finance-focused feed" line in the 2026-09-20 `task=all` verification entry in `EXPERIMENTS.md` is wrong and should be read against this ticket.

**Comment by imanueldrexel** [comment]
**Product decision 2026-09-20 (product owner) -- how this ticket is being split**

This ticket's fixes (1), (2) and (4) stand and **ship first**, unchanged: the symmetric `FINANCE_CATEGORIES` allow-list, the re-queue of the ~1,100 finance-category `rejected_filter` rows, and the per-run pass-rate stat. The allow-list is not a keyword-on-text rule -- it is the publisher's own editorial classification, which is the most reliable signal the pipeline has, and it is what makes the rest of this plan tractable: after it lands, the semantic score only ever arbitrates the genuinely mixed buckets (`news`, `berita`, `news-sitemap`, regional subdomains).

Fix (3) -- "replace the phrase anchors with sentence-length descriptions and re-tune the threshold" -- is **superseded**. Three tuning passes on the cosine-vs-phrases signal (0.35 threshold, SYS-31 deny-list, SYS-31 margin) have each been found wanting by the next audit; a fourth pass on the same signal is not the fix. The scoring step is being replaced outright:

- **SYS-36 (#120)** -- written finance/economy rubric + a Gemini-labelled, stratified, *reject-inclusive* gold set (`tests/fixtures/filter_gold.jsonl`). The rubric fixes two scope questions explicitly: economic policy is IN (the Bahlil / Menhub / DJP / Kemenkeu rejects on this ticket), electoral and party politics stay OUT (PRD §1; `OFFTOPIC_INTERESTS` already says so).
- **SYS-37 (#121)** -- a supervised linear probe on the *same* MiniLM embedding the filter already computes replaces threshold + margin in `is_important`. Gate: recall ≥ 0.90 on the finance-section stratum, precision ≥ 0.90 on the mixed buckets, and this ticket's eight worked examples pass through the real code. That is the "recall guard" of fix (4), made into a CI assertion.
- **SYS-38 (#122)** -- re-score all 28,450 `rejected_filter` rows with the probe (headline-only, lenient) and re-queue the passes, so the recall fix reaches the backlog. This ticket's own re-queue (category allow-set) runs first and is a subset.
- **OPS-18 (#123)** -- weekly Gemini-judged sample of 100 saved + 100 rejected, Telegram alert below 90% agreement. The regression this ticket describes ran for a week unnoticed; that is the gap OPS-18 closes.

**Why not an LLM at runtime** (the question that prompted the split): cost is not the constraint -- ~400-1,000 decisions/day (~10M tokens/month) is single-digit dollars on any current cheap model. The constraints are that `rejected_filter` is terminal (SYS-16) and `process_event` swallows `BaseException`, so an API failure in the per-article path has to degrade to a local decision anyway; `tests/unit` is offline by design; and the pipeline already computes the embedding a linear classifier needs, for free. The LLM's real strength -- no training data, good on edge cases -- is a labeller's job (SYS-36) and a judge's job (OPS-18). A runtime LLM tie-break for the probe's uncertain band stays possible and will be decided on SYS-37's measured band size, not filed now.

Done-when for **this** ticket is unchanged except that the "eight worked examples pass" condition transfers to SYS-37 (it cannot be met honestly by the allow-list alone -- those rows pass because of their category, not because the score recognises them).

**Comment by imanueldrexel** [comment]
**Scope change 2026-09-20 (owner decision, same day as the comment above):** the product now keeps everything **news-worthy**, not only finance/economy -- rubric on SYS-36 (#120): hard news in (politics, law/crime, disasters, international, ...), soft content out (recipes, `doa`, horoscopes, gossip, listicles, fiction, **sports**). What it changes on *this* ticket, which still ships first:

- Fix (1), the symmetric category gate: the allow-set widens from finance sections to **every hard-news section** (`news`, `nasional`, `politik`, `hukum`, `internasional`/`dunia`, `daerah`/`regional`, `megapolitan`, ... plus the finance ones), and **`kriminal` leaves the deny-list** (`OFFTOPIC_CATEGORIES_EXACT` and `OFFTOPIC_CATEGORY_KEYWORDS_SUBSTRING`) -- crime is news. `olahraga`/`bola`/`sport` stay denied (sports out).
- Fix (2), the re-queue: still the finance-category rows only (~1,100) -- the general-section rejects are SYS-38's (#122) job once the probe exists, because on `news`/`berita` the category alone does not say news-worthy.
- Fix (4), the pass-rate stat: measured per allow-set category as before.
- Fix (3) stays superseded by SYS-37 (#121), which is now a multi-class probe (news-worthy gate + `articles.topic`).

Downstream consequences are their own tickets: SRC-34 (#124) crawler widening, SYS-39 (#125) extraction capacity, SYS-40 (#126) newsletter by topic + PRD rewrite, SYS-41 (#127) vector-store capacity.

**Comment by imanueldrexel** [comment]
**Fixes (1), (2 — script + dry-run), (4) landed 2026-09-20 — commit `6792255`** (with SYS-34 #115, same file). Fix (3) superseded by SYS-37 as decided. **Waiting on the owner for the re-queue.**

**Fix (1), symmetric gate.** `category_gate()` in `core/importance_filter.py` is now the one category decision: `deny` (soft vertical) → rejected before the score; `allow` (`NEWS_CATEGORIES` = `NEWS_CATEGORIES_FINANCE` ∪ `NEWS_CATEGORIES_HARD`) → kept before the score; `None` → the score decides. Exact match on the section name (or the last `/` segment of a host-qualified key: `bali/politik`, `finance/keuangan`). `kriminal` left the deny-list (and `edukasi` — see SYS-34). `SemanticImportanceFilter.classify()` returns `(is_important, gate, score_passed, scores)`; `is_important` is a wrapper; `backfill_sitemap_crawl_status.py`'s `classify_batch` goes through the same `category_gate` (it had been rejecting finance rows by headline score exactly as the live filter did). `INTERESTS`, 0.35, `OFFTOPIC_INTERESTS`, margin untouched.

*Deviation from the 2026-09-20 comment, for the owner to overrule:* **`news`/`berita`/`news-sitemap` are not on the allow-set.** The comment lists `news` but also says "on `news`/`berita` the category alone does not say news-worthy"; a random 4-title sample of KUMPARAN `news` was 1 hard news + a photo gallery + a how-to + an opinion piece, VIVA `news-sitemap` 0 of 4; and SYS-37's precision gate is measured on exactly those buckets — a short-circuit there would make it vacuous. Every other name on both lists was decided from a 3-4 title sample (52 pairs sampled; table in `EXPERIMENTS.md`, 2026-09-20 Phase 0 entry): `pendidikan`, `syariah`, `research`, `mymoney`, `csr`, `kesehatan`, `otomotif`, `humaniora` sampled mixed and stay with the score. One-line change if you want `news` in.

**Fix (2), re-queue — pending owner.** `venv\Scripts\python.exe scripts/requeue_finance_rejects.py --dry-run` → **2,080** `rejected_filter` rows under finance categories (the ~1,100 estimate used a narrower list; WARTAEKONOMI's `digital_economy` 150, `energi` 147, `industri` 126, `asuransi` 80, `infrastruktur` 62, ... add ~900). Top: MEDIAINDONESIA `ekonomi` 179, TVONENEWS `ekonomi` 127, LIPUTAN6 `bisnis` 111, WARTAEKONOMI `bisnis` 104, SUARA `bisnis` 94, TIRTO `ekonomi` 92, SINDONEWS `ekbis` 76, TEMPO `ekonomi` 74. The real run backs the rows up, sets `pending` **and clears `last_crawl_attempt`** (they were stamped this week; the 7-day cooldown would otherwise hide them for a week) in one transaction. Nothing is fetched by the script; the rows re-enter the queue for the next `task=full_text` on their source — allowed today only on the eight finance-narrowed sources per the HANDOFF rules; the rest wait for SYS-37. `--keep-cooldown` if you prefer them to wait.

**Fix (4), per-run stat.** Every `task=full_text` logs `filter decisions this run: allow=… deny=… score=… empty=…` and, per news category, `semantic-score recall on news categories … bisnis=62/100 (62%)` — what the score *alone* would have kept among articles the category saved. This is the recall number the ticket was filed on and will read ~50% until SYS-37; INFO, not a WARNING that fires every run. Also: an article with an empty body is now `empty_extraction` *before* the filter (under the allow-set a bare title would have been waved through and chunked to nothing).

Done-when as amended by the comment:
- [x] (transferred to SYS-37) the eight worked examples pass through the real `is_important` — they do, via `gate=allow` with their real body heads (`tests/unit/test_importance_filter.py::test_sys35_worked_examples_pass_under_their_finance_category`, real model); the score half is SYS-37's.
- [ ] a real `task=full_text` over a finance-section backlog saves ≥ 90% of extracted rows — after the re-queue; LIPUTAN6 `bisnis` proposed (111 re-queued rows).
- [ ] the ~1,100 (now 2,080) rows re-queued and re-crawled — pending owner.
- [x] the pass-rate stat exists.

Tests: `tests/unit` 634 → 655, `tests/integration` 102. Mutation checks: removing the allow short-circuit fails 2 tests; putting `kriminal` back on the deny-list fails at import (the two vocabularies are asserted disjoint).

What I would not trust yet: post-SYS-37, an allow-set short-circuit means the probe never gets to reject a listicle inside `nasional`; SYS-37 should decide whether `allow` means "never reject" or "never reject *and* still compute the topic".

**Comment by imanueldrexel** [comment]
**Re-queue executed and the recall verified live, 2026-09-20 15:21 UTC (owner approval in session).**
- `requeue_finance_rejects.py`: **2,105** rows (the 2,080 + CNN's 25 `ekonomi` rejects, visible only after SRC-36's category rewrite) → `pending`, cooldown cleared; backup `backups/sys35_requeue_finance_rejects_20260920T152137Z.json`. `finance-category rows still rejected_filter` → **0**.
- Real `task=full_text` LIPUTAN6 (queue = exactly the 111 re-queued `bisnis` rows): `attempted=111 extracted=111 passed_filter=111 saved=111` — **100%** (Done-when ≥ 90%). The new stat line read `filter decisions this run: allow=111` and `semantic-score recall on news categories … bisnis=0/111 (0%)` — correct, not a bug: these were precisely the rows the score had rejected, and it rejects them again; that is the recall loss this ticket was filed on, now printed per run.
- Same on TIRTO (`95/95 saved`) and IDNTIMES `business` (46 of 57; 11 fetch failures). Rows re-queued on the non-eight sources (MEDIAINDONESIA, WARTAEKONOMI, TVONENEWS, …) stay `pending` until their `task=full_text` is allowed by the HANDOFF rules.

All Done-when items met (the score half of "the eight examples pass" is SYS-37's, per the owner's comment). Closing.

**Comment by imanueldrexel** [comment]
Closing: gate symmetric, 2,105 rows re-queued, LIPUTAN6 bisnis backlog saved 111/111, pass-rate stat live.

**Comment by imanueldrexel** [comment]
**Owner decision 2026-09-20 (in session):** `news` (and `berita`/`news-sitemap`/`sitemap`) is **not** on the always-keep allow-set. Those buckets go to the classifier — today the cosine score, after SYS-37 the probe — like every other mixed bucket. Rationale accepted: KUMPARAN `news` sampled 1-of-4 hard news, VIVA `news-sitemap` 0-of-4, and SYS-37's precision gate is measured on exactly these buckets. This supersedes the `news` entry in the 2026-09-20 scope-change comment above; `NEWS_CATEGORIES` in `core/importance_filter.py` and `TestNewsCategoryAllowSet::test_mixed_buckets_are_neither` pin it.
